### PR TITLE
Yandex S3 output

### PR DESCRIPTION
**Pull request by nar3k**
- Added Yandex S3 output
- fixed GoLint error

**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area config

/area outputs

> /area tests

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/falcosidekick/issues/258

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -47,6 +47,7 @@ var (
 	wavefrontClient     *outputs.Client
 	fissionClient       *outputs.Client
 	grafanaClient       *outputs.Client
+	yandexS3Client      *outputs.Client
```
Comment (Issif) [inline main.go:50]:
```suggestion
	yandexClient      *outputs.Client
```

The client will be used for both S3 and Log

File: main.go [hunk 2]
```diff
@@ -450,6 +451,20 @@ func init() {
 		}
 		log.Printf("[INFO]  : Enabled Outputs : %s\n", outputs.EnabledOutputs)
 	}
+
+	if config.Yandex.S3.Bucket != "" {
+		var err error
+		yandexS3Client, err = outputs.NewYandexS3Client(config, stats, promStats, statsdClient, dogstatsdClient)
```
Comment (Issif) [inline main.go:457]:
```suggestion
		yandexClient, err = outputs.NewYandexClient(config, stats, promStats, statsdClient, dogstatsdClient)
```

See my previous comment

File: main.go [hunk 3]
```diff
@@ -450,6 +451,20 @@ func init() {
...
+		if err != nil {
+			log.Printf("[ERROR] : Yandex - %v\n", err)
```
Comment (Issif) [inline main.go:459]:
```suggestion
            config.Yandex.S3.Bucket = ""
			log.Printf("[ERROR] : Yandex - %v\n", err)
```

File: outputs/yandex.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"github.com/falcosecurity/falcosidekick/types"
+)
+
+// NewAWSClient returns a new output.Client for accessing the AWS API.
+func NewYandexS3Client(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
```
Comment (Issif) [inline outputs/yandex.go:21]:
```suggestion
// NewYandexClient returns a new output.Client for accessing the Yandex API.
func NewYandexClient(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
```

File: outputs/yandex.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+	if config.Yandex.AccessKeyID != "" && config.Yandex.SecretAccessKey != "" {
+		err1 := os.Setenv("AWS_ACCESS_KEY_ID", config.Yandex.AccessKeyID)
+		err2 := os.Setenv("AWS_SECRET_ACCESS_KEY", config.Yandex.SecretAccessKey)
```
Comment (Issif) [inline outputs/yandex.go:25]: if both AWS and Yandex are used, there will be a conflict. Is Yandex also able to use different keys?

File: outputs/yandex.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		if err1 != nil || err2 != nil {
+			log.Printf("[ERROR] : AWS - Error setting AWS env vars")
```
Comment (Issif) [inline outputs/yandex.go:27]:
```suggestion
			log.Printf("[ERROR] : Yandex - Error setting Yandex env vars")
```

File: outputs/yandex.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			return nil, errors.New("Error setting AWS env vars")
```
Comment (Issif) [inline outputs/yandex.go:28]:
```suggestion
			return nil, errors.New("Error setting Yandex env vars")
```

File: outputs/yandex.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	sess, err := session.NewSession(&aws.Config{
+		Region:   aws.String(config.Yandex.Region),
+		Endpoint: aws.String(config.Yandex.Endpoint)})
+	if err != nil {
+		log.Printf("[ERROR] : AWS - %v\n", "Error while creating AWS Session")
+		return nil, errors.New("Error while creating AWS Session")
+	}
+
+	log.Printf("[INFO] : Yandex S3 session has been configured successfully")
```
Comment (Issif) [inline outputs/yandex.go:39]:
```suggestion
```

Not useful as lack of error message means it's ok

File: outputs/yandex.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+			log.Printf("[ERROR] : AWS - Error setting AWS env vars")
+			return nil, errors.New("Error setting AWS env vars")
+		}
+	}
+	sess, err := session.NewSession(&aws.Config{
+		Region:   aws.String(config.Yandex.Region),
+		Endpoint: aws.String(config.Yandex.Endpoint)})
+	if err != nil {
+		log.Printf("[ERROR] : AWS - %v\n", "Error while creating AWS Session")
+		return nil, errors.New("Error while creating AWS Session")
```
Comment (Issif) [inline outputs/yandex.go:36]:
```suggestion
		log.Printf("[ERROR] : AWS - %v\n", "Error while creating Yandex Session")
		return nil, errors.New("Error while creating Yandex Session")
```

File: outputs/yandex.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	log.Printf("[INFO] : Yandex S3 session has been configured successfully")
+
+	return &Client{
+		OutputType:      "YandexS3",
```
Comment (Issif) [inline outputs/yandex.go:42]:
```suggestion
		OutputType:      "Yandex",
```

File: outputs/yandex.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+		Config:          config,
+		AWSSession:      sess,
+		Stats:           stats,
+		PromStats:       promStats,
+		StatsdClient:    statsdClient,
+		DogstatsdClient: dogstatsdClient,
+	}, nil
+}
+
+// UploadS3 upload payload to S3
```
Comment (Issif) [inline outputs/yandex.go:52]:
```suggestion
// UploadYandexS3 uploads payload to Yandex S3
```

File: README.md [hunk 12]
```diff
@@ -406,6 +407,15 @@ grafana:
 
 webui:
   url: "" # WebUI URL, if not empty, WebUI output is enabled
+
+yandex:
+  aws:
+  # accesskeyid: "" # aws access key (optional if you use EC2 Instance Profile)
+  # secretaccesskey: "" # aws secret access key (optional if you use EC2 Instance Profile)
+  s3:
+    # bucket: "falcosidekick" # AWS S3, bucket name
```
Comment (Issif) [inline README.md:416]:
```suggestion
    # bucket: "falcosidekick" # Yandex S3, bucket name
```

File: main.go [hunk 13]
```diff
@@ -450,6 +451,20 @@ func init() {
...
+		} else {
+			if config.Yandex.S3.Bucket != "" {
+				outputs.EnabledOutputs = append(outputs.EnabledOutputs, "YandexS3Client")
```
Comment (Issif) [inline main.go:462]:
```suggestion
				outputs.EnabledOutputs = append(outputs.EnabledOutputs, "YandexS3")
```

File: main.go [hunk 14]
```diff
@@ -450,6 +451,20 @@ func init() {
...
+			}
+		}
+		log.Printf("[INFO]  : Enabled Outputs : %s\n", outputs.EnabledOutputs)
+	}
```
Comment (Issif) [inline main.go:466]:
```suggestion
    }
	log.Printf("[INFO]  : Enabled Outputs : %s\n", outputs.EnabledOutputs)
```

File: main.go [hunk 15]
```diff
@@ -450,6 +451,20 @@ func init() {
 		}
 		log.Printf("[INFO]  : Enabled Outputs : %s\n", outputs.EnabledOutputs)
```
Comment (Issif) [inline main.go:452]:
```suggestion
		log.Printf("[INFO]  : Enabled Outputs : %s\n", outputs.EnabledOutputs)
```

My bad, this line should not be there

File: outputs/yandex.go [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	key := fmt.Sprintf("%s/%s/%s.json", prefix, t.Format("2006-01-02"), t.Format(time.RFC3339Nano))
+	_, err := s3.New(c.AWSSession).PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(c.Config.Yandex.S3.Bucket),
+		Key:    aws.String(key),
+		Body:   bytes.NewReader(f),
+	})
+	if err != nil {
+		go c.CountMetric("outputs", 1, []string{"output:awss3", "status:error"})
+		c.PromStats.Outputs.With(map[string]string{"destination": "awss3", "status": Error}).Inc()
```
Comment (Issif) [inline outputs/yandex.go:68]:
```suggestion
		go c.CountMetric("outputs", 1, []string{"output:yandexs3", "status:error"})
		c.PromStats.Outputs.With(map[string]string{"destination": "yandexs3", "status": Error}).Inc()
```

File: outputs/yandex.go [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+		log.Printf("[ERROR] : %v S3 - %v\n", c.OutputType, err.Error())
+		return
+	}
+
+	log.Printf("[INFO]  : %v S3 - Upload payload OK\n", c.OutputType)
+
+	go c.CountMetric("outputs", 1, []string{"output:awss3", "status:ok"})
+	c.PromStats.Outputs.With(map[string]string{"destination": "awss3", "status": "ok"}).Inc()
```
Comment (Issif) [inline outputs/yandex.go:76]:
```suggestion
	go c.CountMetric("outputs", 1, []string{"output:yandexs3", "status:ok"})
	c.PromStats.Outputs.With(map[string]string{"destination": "yandexs3", "status": "ok"}).Inc()
```

**Review by Issif — CHANGES_REQUESTED** [review]

File: README.md [hunk 18]
```diff
@@ -406,6 +407,16 @@ grafana:
...
+
+yandex:
+  # accesskeyid: "" # yandex access key
+  # secretaccesskey: "" # yandex secret access key
+  # region : "" # yandex region (default: ru-central-1)
+  s3:
+    #endpoint: "" # yandex s3 endpoint (default: ru-central-1)
+    # bucket: "falcosidekick" # Yandex S3, bucket name
+    # prefix : "" # name of prefix, keys will have format: s3://<bucket>/<prefix>/YYYY-MM-DD/YYYY-MM-DDTHH:mm:ss.s+01:00.json
+    # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|erro
```
Comment (Issif) [inline README.md:419]: OK for me, can you also add these values into env vars section of README. please

File: config.go [hunk 19]
```diff
@@ -358,6 +366,7 @@ func getConfig() *types.Configuration {
 	c.Fission.MinimumPriority = checkPriority(c.Fission.MinimumPriority)
 	c.Rabbitmq.MinimumPriority = checkPriority(c.Rabbitmq.MinimumPriority)
 	c.Wavefront.MinimumPriority = checkPriority(c.Wavefront.MinimumPriority)
+	c.Yandex.S3.MinimumPriority = checkPriority(c.Wavefront.MinimumPriority)
```
Comment (Issif) [inline config.go:369]:
```suggestion
	c.Yandex.S3.MinimumPriority = checkPriority(c.Yandex.S3.MinimumPriority)
```

File: main.go [hunk 20]
```diff
@@ -47,6 +47,7 @@ var (
 	wavefrontClient     *outputs.Client
 	fissionClient       *outputs.Client
 	grafanaClient       *outputs.Client
+	yandexClient      *outputs.Client
```
Comment (Issif) [inline main.go:50]:
```suggestion
	yandexClient        *outputs.Client
```

File: outputs/yandex.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+			log.Println("create file: ", err)
+
+		}
+		err = t.Execute(file, p)
+		if err != nil {
+			log.Print("execute: ", err)
+
+		}
+
+	}
```
Comment (Issif) [inline outputs/yandex.go:52]: I don't understand the purpose of this, why are you creating an `.aws/config` file? Can't you use env vars like for AWS Client?
Please, do not write anything on disk, keep everything in memory.

If it's just for avoiding a conflict with AWS credentials aside, you can create use this function for setting `Credentials`: [credential.NewStaticCredentials](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#NewStaticCredentials)

**Review by Issif — CHANGES_REQUESTED** [review]

**Comment by nar3k** [comment]
I hope it should be fine right now

Sorry for the spam , im new to github community - this PR gave me a lot of experience

**Comment by Issif** [comment]
> I hope it should be fine right now
> 
> Sorry for the spam , im new to github community - this PR gave me a lot of experience

No problem, I'm here for helping you to contribute.

File: README.md [hunk 22]
```diff
@@ -755,6 +766,14 @@ care of lower/uppercases**) : `yaml: a.b --> envvar: A_B` :
 - **GRAFANA_CHECKCERT**: check if ssl certificate of the output is valid (default: true)
 - **GRAFANA_MINIMUMPRIORITY**: minimum priority of event for using this output, order is 
   `emergency|alert|critical|error|warning|notice|informational|debug or "" (default)`
+- **YANDEX_ACCESSKEYID** : Yandex Access Key Id 
+- **YANDEX_SECRETACCESSKEY** : AWS Secret Access Key 
+- **YANDEX_REGION**: Yandex  region (default: ru-central-1)
+- **YANDEX_S3_ENDPOINT**: yandex storage endpoint (default: https://storage.yandexcloud.net)
```
Comment (Issif) [inline README.md:772]:
```suggestion
- **YANDEX_SECRETACCESSKEY** : Yandex Secret Access Key 
- **YANDEX_REGION**: Yandex region (default: ru-central-1)
- **YANDEX_S3_ENDPOINT**: Yandex storage endpoint (default: https://storage.yandexcloud.net) 
```

**Review by Issif — CHANGES_REQUESTED** [review]

**Review by Issif — APPROVED** [review]
well done and congrats 👍

**Pull request merged**